[4.0] RavenDB-10581 when we are using / in the ID, then we need to check if the generated ID does not overlap already existing one

ppekrol opened the pull request.

ppekrol committed:
RavenDB-10581 when we are using / in the ID, then we need to check if the generated ID does not overlap already existing one

ppekrol committed:
RavenDB-10581 introduced OverlappingGeneratedNewIdException

ayende commented on src/Raven.Server/Documents/TransactionOperationsMerger.cs:
```diff
@@ -60,7 +62,12 @@ public TransactionOperationsMerger(DocumentDatabase parent, CancellationToken sh
+
+        public abstract class DocumentPutTransactionCommand : MergedTransactionCommand
+        {
+            public bool CheckIfGeneratedIdIsNotOverlapping;
```
Why, this is relevant only on PUT on new docs. Not relevant anywhere else (`PATCH`, for example)

ayende commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -301,11 +316,31 @@ private string BuildDocumentId(string id, long newEtag, out bool knownNewId)
+                            Format.Backwards.WriteNumber(valuePtr + i - 1, (ulong)newEtag);
+                        else
+                        {
+                            while (true)
```
I don't want this solution, this does a scan, and can get into bad situation.
Just change the generated doc id in this case.

ayende commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -170,7 +172,17 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+                        catch (VoronConcurrencyErrorException)
+                        {
+                            if (knownNewId && checkIfGeneratedIdIsNotOverlapping == false)
+                                ThrowOverlappingGeneratedNewIdException(id);
```
This here should recurse immediately and try (once).

ayende committed:
RavenDB-10581
 - Moved the state of retries to the calling commands.
 - Command can now ask to be retried (once)
 - Will work even on two commands where the first failed and the second also have issues.

ayende committed:
RavenDB-10581 - Reverting other changes, avoid infinite loop on error, removed unused exception type

ayende committed:
RavenDB-10581 - Reverting all changes to DocumentPutAction

ayende committed:
RavenDB-10581 Avoid repeating myself and ensuring that we use the same format as expected of us

ppekrol commented:
LGTM

arekpalinski commented:
LGTM, although " Server compilation failed. "

ppekrol commented:
I will fix that

ppekrol committed:
RavenDB-10581 fixing compilation
